src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -0,0 +1,406 @@
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Cache manager for resources.
+ * <p>
+ * This class handles caching the resource listing calls on a pack, pack type and namespace level.
+ */
+public class ResourceCacheManager
+{
+
+    /**
+     * Indicates if the underlying namespaced managers need to support reloading.
+     * Disabled for the (in-jar / downloaded) default vanilla pack.
+     */
+    private final boolean supportsReloading;
+    /**
+     * Indicates if the indexing of the file-tree should happen off-thread or on-thread and block the process accordingly.
+     */
+    private final ForgeConfigSpec.BooleanValue indexOffThreadConfigOption;
+
+    /**
+     * The path builder (different users have different requirements for how they want to handle this)
+     */
+    private final BiFunction<PackType, String, Path> pathBuilder;
+    /**
+     * The individual sub managers by pack type and namespace
+     */
+    private final Map<PackTypeAndNamespace, NamespacedResourceCacheManager> managersByNamespace = Maps.newConcurrentMap();
+
+    /**
+     * Creates a new instance of a resource cache manager.
+     *
+     * @param supportsReloading    True to make the namespace specific managers support reloading the cache.
+     * @param indexOffThreadConfig True to index the file-tree off-thread.
+     * @param pathBuilder          The path builder to use.
+     */
+    public ResourceCacheManager(final boolean supportsReloading, final ForgeConfigSpec.BooleanValue indexOffThreadConfig, final BiFunction<PackType, String, Path> pathBuilder)
+    {
+        this.supportsReloading = supportsReloading;
+        this.indexOffThreadConfigOption = indexOffThreadConfig;
+        this.pathBuilder = pathBuilder;
+    }
+
+    public boolean shouldIndexOffThread()
+    {
+        return !getConfigValue(this.indexOffThreadConfigOption);
+    }
+
+    public static boolean shouldUseCache()
+    {
+        return getConfigValue(ForgeConfig.COMMON.cachePackAccess);
+    }
+
+    private static boolean getConfigValue(Supplier<Boolean> configValue)
+    {
+        // Yes we catch the early loading error on purpose.
+        // Both the indexing off-thread and if caching is enabled are configurable, so we need to catch the error.
+        // So when they are loaded early we preserve the default behaviour which is to not cache or index on thread.
```

Comment:
The problem is that pack size plays a role here in the performance gain. There are constructs where the mod asset count vs the vanilla asset count needs to be a bit balance. Yes in most situations it makes sense to them all off-thread. But in a vanilla like scenario where mods are in play which have virtually no assets there are some benefits to be gained from performing mod indexing on thread. The same can be set for vanilla. If you play in a vanilla only environment it might make sense to do the init on thread, wait for it and then completely use in cache only resources.

So I decided to handle it this way. It provides greater flexibility in situations where it makes sense.